src/main/java/org/candlepin/subscriptions/capacity/CapacityReconciliationController.java

```diff
@@ -0,0 +1,118 @@
+
+  private final OfferingRepository offeringRepository;
+  private final ProductWhitelist productWhitelist;
+  private final CapacityProductExtractor productExtractor;
+  private final SubscriptionCapacityRepository subscriptionCapacityRepository;
+  private final Counter capacityRecordsCreated;
+  private final Counter capacityRecordsUpdated;
+  private final Counter capacityRecordsDeleted;
+
+  public CapacityReconciliationController(
+      OfferingRepository offeringRepository,
+      ProductWhitelist productWhitelist,
+      CapacityProductExtractor productExtractor,
+      SubscriptionCapacityRepository subscriptionCapacityRepository,
+      MeterRegistry meterRegistry) {
+    this.offeringRepository = offeringRepository;
+    this.productWhitelist = productWhitelist;
+    this.productExtractor = productExtractor;
+    this.subscriptionCapacityRepository = subscriptionCapacityRepository;
+    capacityRecordsCreated = meterRegistry.counter("rhsm-subscriptions.capacity.records_created");
+    capacityRecordsUpdated = meterRegistry.counter("rhsm-subscriptions.capacity.records_updated");
+    capacityRecordsDeleted = meterRegistry.counter("rhsm-subscriptions.capacity.records_deleted");
+  }
+
+  public void reconcileCapacityForSubscription(Subscription subscription) {
+
+    Collection<SubscriptionCapacity> newCapacities = mapSubscriptionToCapacities(subscription);
+    reconcileSubscriptionCapacities(
+        newCapacities, subscription.getSubscriptionId(), subscription.getSku());
+  }
+
+  private Collection<SubscriptionCapacity> mapSubscriptionToCapacities(Subscription subscription) {
+
+    Offering offering = offeringRepository.getById(subscription.getSku());
+    Set<String> products =
+        productExtractor.getProducts(
+            offering.getProductIds().stream().map(Object::toString).collect(Collectors.toSet()));
+
+    return products.stream()
+        .map(product -> from(subscription, offering, product))
+        .collect(Collectors.toList());
+  }
+
+  private void reconcileSubscriptionCapacities(
+      Collection<SubscriptionCapacity> newCapacities, String subscriptionId, String sku) {
+
+    Collection<SubscriptionCapacity> toSave = new ArrayList<>();
+    Map<SubscriptionCapacityKey, SubscriptionCapacity> existingCapacityMap =
+        subscriptionCapacityRepository.findByKeySubscriptionId(subscriptionId).stream()
+            .collect(Collectors.toMap(SubscriptionCapacity::getKey, Function.identity()));
+
+    if (productWhitelist.productIdMatches(sku)) {
+      newCapacities.forEach(
+          newCapacity -> {
+            toSave.add(newCapacity);
+            SubscriptionCapacity oldVersion = existingCapacityMap.remove(newCapacity.getKey());
+            if (oldVersion != null) {
+              capacityRecordsUpdated.increment();
+            } else {
+              capacityRecordsCreated.increment();
+            }
+          });
+      subscriptionCapacityRepository.saveAll(toSave);
```

Comment:
Do you know what happens in the event that this `saveAll` throws an exception?  Where I'm going with this is that I'm wondering if we need a `@Transactional` on this method or not, to "revert" the `capacityRecordsUpdated` and `capacityRecordsCreated` values that have already been incremented.  

Alternatively wondering if we might able to leverage java stream's `groupingBy` collector to be like "Ok here's a list of the capacities that existed in the database before, here's another list of the capacities that didn't.  Save all the capacities, then just count the number that fit either category.

```java
var dividedCapacities = newCapacities.groupingByMagic();

subscriptionCapacityRepository.saveAll(newCapacities);

// i think that grouping by gives some obnoxious map, so this pseudocode about arrays isn't even a little accurate on how to actually do it
capacityRecordsUpdated.increment(dividedCapacities[0].size());
capacityRecordsCreated.increment(dividedCapacities[1].size());

```

Reply:
`reconcileCapacityForSubscription` is only ever called by `syncSubscription` which is executed within a transaction, but it is also a public method, so I added `@Transactional` to capacity reconciliation as well to ensure it always executes within a transaction.